### PR TITLE
Handle private gateways more reliably

### PR DESCRIPTION
**Pull request by remibergsma**
When initialising a VPC router we need to know which IP/device corresponds to a private gateway. This is to solve a problem when stop/starting a VPC router (which gets the private gateway config as a guest network and as a result breaks the functionality). You read it right, the private gateway is sent as type=guest after reboot and type=public initially. 

Before this change, you could add a private gw to a running router but you couldn't restart it (it would mix up the tiers). Now the private gateway is detected properly and it works just fine.

Booting without private gateway:
```
root@r-167-VM:~# cat /etc/cloudstack/cmdline.json 
{
    "config": {
        "baremetalnotificationapikey": "V2l1u3wKJVan01h8kq63-5Y5Ia3VLEW1v_Z6i-31QIRJXlt5vkqaqf6DVcdK0jP3u79SW6X9pqJSLSwQP2c2Rw", 
        "baremetalnotificationsecuritykey": "OXI16srCrxFBi-xOtEwcYqwLlMfSFTlTg66YHtXBBqR7HNN1us3HP5zWOKxfVmz4a3C1kUNLPrUH13gNmZlu4w", 
        "disable_rp_filter": "true", 
        "dns1": "8.8.8.8", 
        "domain": "cs2cloud", 
        "eth0ip": "169.254.0.42", 
        "eth0mask": "255.255.0.0", 
        "host": "192.168.22.61", 
        "name": "r-167-VM", 
        "port": "8080", 
        "privategateway": "None", 
        "redundant_router": "false", 
        "template": "domP", 
        "type": "vpcrouter", 
        "vpccidr": "10.0.0.0/24"
    }, 
    "id": "cmdline"
```

Booting with private gateway:
```
root@r-167-VM:~# cat /etc/cloudstack/cmdline.json 
{
    "config": {
        "baremetalnotificationapikey": "V2l1u3wKJVan01h8kq63-5Y5Ia3VLEW1v_Z6i-31QIRJXlt5vkqaqf6DVcdK0jP3u79SW6X9pqJSLSwQP2c2Rw", 
        "baremetalnotificationsecuritykey": "OXI16srCrxFBi-xOtEwcYqwLlMfSFTlTg66YHtXBBqR7HNN1us3HP5zWOKxfVmz4a3C1kUNLPrUH13gNmZlu4w", 
        "disable_rp_filter": "true", 
        "dns1": "8.8.8.8", 
        "domain": "cs2cloud", 
        "eth0ip": "169.254.2.227", 
        "eth0mask": "255.255.0.0", 
        "host": "192.168.22.61", 
        "name": "r-167-VM", 
        "port": "8080", 
        "privategateway": "10.201.10.1", 
        "redundant_router": "false", 
        "template": "domP", 
        "type": "vpcrouter", 
        "vpccidr": "10.0.0.0/24"
    }, 
    "id": "cmdline"
```

And:
```
cat cmdline 
vpccidr=10.0.0.0/24 domain=cs2cloud dns1=8.8.8.8 privategateway=10.201.10.1 template=domP name=r-167-VM eth0ip=169.254.2.227 eth0mask=255.255.0.0 type=vpcrouter disable_rp_filter=true baremetalnotificationsecuritykey=OXI16srCrxFBi-xOtEwcYqwLlMfSFTlTg66YHtXBBqR7HNN1us3HP5zWOKxfVmz4a3C1kUNLPrUH13gNmZlu4w baremetalnotificationapikey=V2l1u3wKJVan01h8kq63-5Y5Ia3VLEW1v_Z6i-31QIRJXlt5vkqaqf6DVcdK0jP3u79SW6X9pqJSLSwQP2c2Rw host=192.168.22.61 port=8080
```

Logs:
```
2016-02-24 20:08:45,723 DEBUG [c.c.n.r.VpcVirtualNetworkApplianceManagerImpl] (Work-Job-Executor-4:ctx-458d4c52 job-1402/job-1403 ctx-d5355fca) (logid:5772906c) Set privategateway field in cmd_line.json to 10.201.10.1
```

### CODE REVIEW

**Comment by swill** [comment]
I will add this to my CI queue.  Do we want to maintain commit history for the two commits in this case?

Also, I am looking for code reviews on this one.  Thanks...

**Comment by DaanHoogland** [comment]
@swill I don't think many people read spurious comments on the github PRs. only fanatics do :p 
I am for safeguarding as much history as possible and only removing the typical to`n'fro experiments.
looking at the code now

File: systemvm/patches/debian/config/opt/cloud/bin/merge.py [hunk 1]
```diff
@@ -274,3 +277,46 @@ def __moveFile(self, origPath, path):
             os.makedirs(path)
         timestamp = str(int(round(time.time())))
         os.rename(origPath, path + "/" + self.fileName + "." + timestamp)
+
+
+class PrivateGatewayHack:
```
Comment (DaanHoogland) [inline systemvm/patches/debian/config/opt/cloud/bin/merge.py]: maybe not the greatest class name ever

File: server/src/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java [hunk 2]
```diff
@@ -260,6 +261,15 @@ public boolean finalizeVirtualMachineProfile(final VirtualMachineProfile profile
...
                 }
+
+                VpcGatewayVO privateGatewayForVpc = _vpcGatewayDao.getPrivateGatewayForVpc(domainRouterVO.getVpcId());
+                if (privateGatewayForVpc != null) {
+                    String ip4Address = privateGatewayForVpc.getIp4Address();
+                    buf.append(" privategateway=").append(ip4Address);
+                    s_logger.debug("Set privategateway field in cmd_line.json to " + ip4Address);
+                } else {
+                    buf.append(" privategateway=None");
+                }
```
Comment (DaanHoogland) [inline server/src/com/cloud/network/router/VpcVirtualNetworkApplianceManagerImpl.java]: matter of style: I would have assigned "None" to ip4Assress and appended the buffer outside the conditional and maybe factorred the block out as method. no issue.

File: systemvm/patches/debian/config/opt/cloud/bin/merge.py [hunk 3]
```diff
@@ -274,3 +277,46 @@ def __moveFile(self, origPath, path):
...
+
+
+    @classmethod
+    def update_network_type_for_privategateway(cls, dbag, data):
+        ip = data['router_guest_ip'] if 'router_guest_ip' in data.keys() else data['public_ip']
+
+        initial_data = cls.load_inital_data()
```
Comment (DaanHoogland) [inline systemvm/patches/debian/config/opt/cloud/bin/merge.py]: Should we have a conditional around load_initial_data? according to lines 148 and 154 this is called twice.

**Comment by DaanHoogland** [comment]
LGTM in spite of comments.

**Comment by kiwiflyer** [comment]
We're pulling this in for testing.

**Comment by swill** [comment]
cool thanks. :)

**Comment by swill** [comment]
### CI RESULTS

```
Tests Run: 85
  Skipped: 0
   Failed: 0
   Errors: 0
 Duration: 4h 32m 05s
```



**Associated Uploads**

**`/tmp/MarvinLogs/DeployDataCenter__May_11_2016_06_54_42_3CPRW6:`**
* [dc_entries.obj](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/DeployDataCenter__May_11_2016_06_54_42_3CPRW6/dc_entries.obj)
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/DeployDataCenter__May_11_2016_06_54_42_3CPRW6/failed_plus_exceptions.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/DeployDataCenter__May_11_2016_06_54_42_3CPRW6/runinfo.txt)

**`/tmp/MarvinLogs/test_network_TS9WMA:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/test_network_TS9WMA/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/test_network_TS9WMA/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/test_network_TS9WMA/runinfo.txt)

**`/tmp/MarvinLogs/test_vpc_routers_M1AWH1:`**
* [failed_plus_exceptions.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/test_vpc_routers_M1AWH1/failed_plus_exceptions.txt)
* [results.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/test_vpc_routers_M1AWH1/results.txt)
* [runinfo.txt](https://objects-east.cloud.ca/v1/e465abe2f9ae4478b9fff416eab61bd9/PR1474/tmp/MarvinLogs/test_vpc_routers_M1AWH1/runinfo.txt)


Uploads will be available until `2016-07-11 02:00:00 +0200 CEST`

*Comment created by [`upr comment`](https://github.com/cloudops/upr).*

**Comment by swill** [comment]
I need one more code review and this one should be ready to merge.  @kiwiflyer did you have a chance to review this one?

**Comment by kiwiflyer** [comment]
LGTM. Tested in a hardware lab. 

tag:mergeready

**Comment by swill** [comment]
Thank you sir.  I will get this merged...

**Pull request merged**